src/wsproxy.ml

```diff
@@ -15,124 +15,115 @@
 
 let get_dir_path () = Printf.sprintf "/var/xapi/"
 
-let ignore_exn t () = Lwt.catch t (fun _ -> Lwt.return_unit)
-
 module LwtWsIteratee = Websockets.Wsprotocol(Lwt)
 open Lwt.Infix
 
+let with_fd = Lwt_support.with_fd
+
 let start path handler =
   let dir_path = get_dir_path () in
   let fd_sock_path = Printf.sprintf "%s%s" dir_path path in
+  Lwt_log.info_f "Starting wsproxy on %s" fd_sock_path >>= fun () ->
   let fd_sock = Lwt_unix.socket Unix.PF_UNIX Unix.SOCK_STREAM 0 in
   Lwt.catch
     (fun () -> Lwt_unix.unlink fd_sock_path)
-    (fun _ -> Lwt.return ()) >>= fun () ->
+    (fun _ -> Lwt.return_unit) >>= fun () ->
   let () = Lwt_unix.bind fd_sock (Unix.ADDR_UNIX fd_sock_path) in
   let () = Lwt_unix.listen fd_sock 5 in
 
   let rec loop () =
+    let ensure_close = function
+      | [] -> Lwt.return_unit
+      | fds ->
+        Lwt_log.warning_f "Closing %d excess fds" (List.length fds) >>= fun () ->
+        Lwt.catch
+          (fun () -> 
+             List.iter (fun fd -> try Unix.close fd with _ -> ()) fds;
+             Lwt.return_unit)
+          (fun _ -> Lwt.return_unit)
+    in
     Lwt.catch
       (fun () ->
-         Lwt_unix.accept fd_sock >>= fun (fd_sock2,_) ->
-         let buffer = String.make 16384 '\000' in
-         Lwt.catch
-           (fun () ->
-              let iov = Lwt_unix.io_vector ~buffer ~offset:0 ~length:16384 in
-              Lwt_unix.recv_msg ~socket:fd_sock2 ~io_vectors:[iov])
-           (fun e ->
-              Printf.printf "Caught exception: %s\n" (Printexc.to_string e);
-              ignore_exn (fun () -> Lwt_unix.close fd_sock2) () >>= fun () ->
-              Lwt.return (0,[]))
-         >>= fun (len,newfds) ->
-         Lwt_unix.close fd_sock2 >>= fun () ->
-         Lwt.catch
-           (fun () ->
-              let fdh :: fdt = newfds in
-              let msg = String.sub buffer 0 len in
-              let fdno = Printf.sprintf "%d" (Obj.magic fdh) in
-              List.iter (fun fd -> Printf.printf "closing fds: %d\n%!" (Obj.magic fd)) fdt;
-              List.iter (fun fd -> try Unix.close fd with _ -> ()) fdt;
-              Printf.printf "About to fixup: %s\n%!" fdno;
-              let fd = Lwt_unix.of_unix_file_descr fdh in
-              Lwt_unix.setsockopt fd Lwt_unix.SO_KEEPALIVE true;
-              handler fd msg)
-           (fun e ->
-              Printf.printf "Caught exception: %s\n" (Printexc.to_string e);
-              List.iter (fun fd -> Printf.printf "closing fd: %d\n%!" (Obj.magic fd)) newfds;
-              List.iter (fun fd -> try Unix.close fd with _ -> ()) newfds;
-              Lwt.return ()) 
-         >>= fun _ ->
-         loop ()
-      )
+         Lwt_unix.accept fd_sock
+         >>= fun (fd_sock',_) ->
+         (* Background thread per connection *)
+         let _ =
+           let buffer = String.make 16384 '\000' in
```

Comment:
Could we move this `String.make` inside `with_fd`? This can also throw exceptions :S 
```
 val String.make int -> char -> string                                                                                                                               
    [String.make n c] returns a fresh string of length [n],                                                                                                           
    filled with the character [c].                                                                                                                                    
                                                                                                                                                                      
    Raise [Invalid_argument] if [n < 0] or [n > ]{!Sys.max_string_length}. 
```

Reply:
This is a good point, however the buffer is used also outside that `with_fd`.
I could modify the functions to pipe the buffer through, but this would make it less readable.
I think we should leave it there: if we fail to create a 16k string, it means that we have serious memory pressure and I would bet that other things will fail even more badly (making the leak of `fd_sock'` not that relevant...)